### PR TITLE
Internal article linking update feature

### PR DESCRIPTION
**Pull request by balazssimongen**
Adding the capability to update linked synced articles to point to the destinaton's articles instead of the source articles if the article exists in the destination.
With this change Salesforce, ServiceNow and Genesys supported only. Zendesk document update is not available yet

### CODE REVIEW

File: src/utils/link-object-extractor.ts [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+    case 'Paragraph':
+      return extractLinkBlocksFromParagraph(block.paragraph);
+    case 'Text':
+      return extractLinkBlocksFromDocumentText(block.text);
+    case 'Image':
+      return extractLinkBlocksFromImage(block.image);
+    case 'OrderedList':
+    case 'UnorderedList':
+      return extractLinkBlocksFromList(block.list);
+    default:
```
Comment (tmelko-genesys) [inline src/utils/link-object-extractor.ts:227]: case 'Table' is missing.

File: src/utils/link-object-extractor.ts [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+    .flat();
+}
+
+function extractLinkBlocksFromTable(
+  table: DocumentBodyTable | undefined,
+): LinkBlock[] {
+  if (!table) {
+    return [];
+  }
+
```
Comment (tmelko-genesys) [inline src/utils/link-object-extractor.ts:111]: table.properties.caption can also have image and text blocks.

File: src/processor/prefix-external-id.spec.ts [hunk 3]
```diff
@@ -79,20 +81,33 @@ describe('PrefixExternalId', () => {
...
+        entities.forEach((item, index) =>
           expect(item.externalId).toBe(
             `${EXTERNAL_ID_PREFIX}${entityType}-${index + 1}`,
           ),
         );
       },
     );
+
+    const entriesArray = Array.from(result.articleLookupTable?.entries() ?? []);
+    entriesArray.forEach(([key, externalLink], index) =>
```
Comment (petergaras) [inline src/processor/prefix-external-id.spec.ts:106]: Here you should also assert that the array has any items in it. Without that you may verify nothing.

File: src/genesys/content-mapper.ts [hunk 4]
```diff
@@ -47,3 +49,26 @@ function documentMapper(article: Document): Document {
...
 }
+
+function buildArticleLookupTable(documents: Document[]) {
+  const lookupTable: Map<string, ExternalLink> = new Map<
+    string,
+    ExternalLink
+  >();
+  documents.forEach((document) => {
+    document.published?.variations.forEach((variation) => {
+      lookupTable.set(`${document.id}#${variation.id}`, {
```
Comment (petergaras) [inline src/genesys/content-mapper.ts:60]: This can be also extracted

File: src/tests/utils/entity-generators.ts [hunk 5]
```diff
@@ -190,6 +250,139 @@ export function generateDocumentWithTable(suffix: string): Document {
...
+              {
+                type: 'Paragraph',
+                paragraph: {
+                  blocks: [
+                    {
+                      type: 'Text',
+                      text: {
+                        text: 'Link 1',
+                        hyperlink:
+                          'https://genesysdev.service-now.com/kb_view.do?sysparm_article=KB0012439',
```
Comment (petergaras) [inline src/tests/utils/entity-generators.ts:274]: Please use a fictive domain

File: src/utils/link-object-extractor.ts [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+  DocumentBodyImage,
+  DocumentBodyList,
+  DocumentBodyTable,
+  DocumentBodyTableCellBlock,
+  DocumentBodyTableRowBlock,
+  DocumentListContentBlock,
+  DocumentTableContentBlock,
+  DocumentText,
+} from 'knowledge-html-converter';
+import { DocumentBodyListBlock } from 'knowledge-html-converter/dist/models/blocks/document-body-list';
```
Comment (petergaras) [inline src/utils/link-object-extractor.ts:18]: Expose this class in html converter, and do not import anything from dep's dist folder.

**Review by tmelko-genesys — APPROVED** [review]

**Review by tkurucz-genesys — APPROVED** [review]

**Review by petergaras — APPROVED** [review]

**Pull request merged**